src/js/node/dgram.ts

```diff
@@ -1,23 +1,1082 @@
+  } else {
+    afterDns(null, null);
+  }
+};
+
+function doSend(ex, self, ip, list, address, port, callback) {
+  const state = self[kStateSymbol];
+
+  if (ex) {
+    if (typeof callback === "function") {
+      process.nextTick(callback, ex);
+      return;
+    }
+
+    process.nextTick(() => self.emit("error", ex));
+    return;
+  }
+  if (!state.handle) {
+    return;
+  }
+  const socket = state.handle.socket;
+  if (!socket) {
+    return;
+  }
+
+  let err = null;
+  let sent = 0;
+  try {
+    if (list.length === 1) {
+      if (port) {
+        socket.send(list[0], port, ip);
+      } else {
+        socket.send(list[0]);
+      }
+      sent = 1;
+    } else {
+      const packets = [];
+      if (port) {
+        for (const buf of list) {
+          packets.push(buf, port, address);
+        }
+      } else {
+        for (const buf of list) {
+          packets.push(buf);
+        }
+      }
+      sent += socket.sendMany(packets);
+    }
+  } catch (e) {
+    err = e;
+  }
+  // TODO check if this makes sense
+  if (callback) {
+    if (err) {
+      process.nextTick(callback, err);
+    } else {
+      process.nextTick(callback, null, sent);
+    }
+  }
+
+  /*
+  const req = new SendWrap();
+  req.list = list;  // Keep reference alive.
+  req.address = address;
+  req.port = port;
+  if (callback) {
+    req.callback = callback;
+    req.oncomplete = afterSend;
+  }
+
+  let err;
+  if (port)
+    err = state.handle.send(req, list, list.length, port, ip, !!callback);
+  else
+    err = state.handle.send(req, list, list.length, !!callback);
+
+  if (err >= 1) {
+    // Synchronous finish. The return code is msg_length + 1 so that we can
+    // distinguish between synchronous success and asynchronous success.
+    if (callback)
+      process.nextTick(callback, null, err - 1);
+    return;
+  }
+
+  if (err && callback) {
+    // Don't emit as error, dgram_legacy.js compatibility
+    const ex = new ExceptionWithHostPort(err, 'send', address, port);
+    process.nextTick(callback, ex);
+  }
+  */
+}
+
+/*
+function afterSend(err, sent) {
+  if (err) {
+    err = new ExceptionWithHostPort(err, 'send', this.address, this.port);
+  } else {
+    err = null;
+  }
+
+  this.callback(err, sent);
+}
+*/
+
+Socket.prototype.close = function (callback) {
+  const state = this[kStateSymbol];
+  const queue = state.queue;
+
+  if (typeof callback === "function") this.on("close", callback);
+
+  if (queue !== undefined) {
+    queue.push(FunctionPrototypeBind(this.close, this));
+    return this;
+  }
+
+  state.receiving = false;
+  state.handle.socket?.close();
+  state.handle.socket = undefined;
+  defaultTriggerAsyncIdScope(this[async_id_symbol], process.nextTick, socketCloseNT, this);
+
+  return this;
+};
+
+Socket.prototype[SymbolAsyncDispose] = async function () {
+  if (!this[kStateSymbol].handle) {
+    return;
+  }
+  console.log(this, this.close, promisify(this.close));
+  return promisify(this.close).$call(this);
```

Comment:
do we need promisify at all?

Reply:
that's what node was doing, and from my reading of the code it makes sense - we want to return a promise from the async function that resolves when this.close is done.
I updated it to just return a new promise that resolves in the close callback